### PR TITLE
Mapeditor refactor, fixes and additional validation checks

### PR DESCRIPTION
**Pull request by MichalZr6**
This PR is made new from https://github.com/vcmi/vcmi/pull/4473 which had messed-up, broken branch.
Commits were cherry-picked.

The PR mainly adds additional validation check when Player is defined (via selecting number of players) but has no heroes or towns assigned which will lead to crash when choosing this map in VCMI client.

### CODE REVIEW

File: mapeditor/mapcontroller.cpp [hunk 1]
```diff
@@ -565,13 +573,13 @@ bool MapController::canPlaceObject(int level, CGObjectInstance * newObj, QString
...
 		auto subTypeName = QString::fromStdString(newObj->subTypeName);
-		error = QString("There can be only one grail object on the map");
+		error = QString("There can only be one grail object on the map.");
 		return false; //maplimit reached
 	}
 	
 	if(defaultPlayer == PlayerColor::NEUTRAL && (newObj->ID == Obj::HERO || newObj->ID == Obj::RANDOM_HERO))
 	{
-		error = "Hero cannot be created as NEUTRAL";
+		error = QString("Hero %1 cannot be created as NEUTRAL.").arg(QString::fromStdString(newObj->instanceName));
```
Comment (kambala-decapitator) [inline mapeditor/mapcontroller.cpp:582]:
```suggestion
		error = QObject::tr("Hero %1 cannot be created as NEUTRAL.").arg(QString::fromStdString(newObj->instanceName));
```

File: mapeditor/mapcontroller.cpp [hunk 2]
```diff
@@ -565,13 +573,13 @@ bool MapController::canPlaceObject(int level, CGObjectInstance * newObj, QString
 	{
 		auto typeName = QString::fromStdString(newObj->typeName);
 		auto subTypeName = QString::fromStdString(newObj->subTypeName);
-		error = QString("There can be only one grail object on the map");
+		error = QString("There can only be one grail object on the map.");
```
Comment (kambala-decapitator) [inline mapeditor/mapcontroller.cpp:576]:
```suggestion
		error = QObject::tr("There can only be one grail object on the map.");
```

File: mapeditor/validator.cpp [hunk 3]
```diff
@@ -58,133 +58,141 @@ std::list<Validator::Issue> Validator::validate(const CMap * map)
...
-						issues.emplace_back(QString(tr("Hero %1 is prohibited by map settings")).arg(ins->type->getNameTranslated().c_str()), false);
+						issues.insert({ tr("Hero %1 is prohibited by map settings").arg(ins->type->getNameTranslated().c_str()), false });
 					
 					if(!allHeroesOnMap.insert(ins->type).second)
-						issues.emplace_back(QString(tr("Hero %1 has duplicate on map")).arg(ins->type->getNameTranslated().c_str()), false);
+						issues.insert({ tr("Hero %1 has duplicate on map").arg(ins->type->getNameTranslated().c_str()), false });
 				}
 				else if(ins->ID != Obj::RANDOM_HERO)
-					issues.emplace_back(QString(tr("Hero %1 has an empty type and must be removed")).arg(ins->instanceName.c_str()), true);
+						issues.insert({ tr("Hero %1 has an empty type and must be removed").arg(ins->instanceName.c_str()), true });
```
Comment (kambala-decapitator) [inline mapeditor/validator.cpp:134]: seems indent level increased by mistake

File: mapeditor/validator.cpp [hunk 4]
```diff
@@ -58,133 +58,141 @@ std::list<Validator::Issue> Validator::validate(const CMap * map)
...
-						if(map->allowedSpells.count(ins->storedArtifact->getScrollSpellID()) == 0)
-							issues.emplace_back(QString(tr("Spell scroll %1 is prohibited by map settings")).arg(ins->storedArtifact->getScrollSpellID().toEntity(VLC->spells())->getNameTranslated().c_str()), false);
-					}
-					else
-						issues.emplace_back(QString(tr("Spell scroll %1 doesn't have instance assigned and must be removed")).arg(ins->instanceName.c_str()), true);
+						if (map->allowedSpells.count(ins->storedArtifact->getScrollSpellID()) == 0)
+							issues.insert({ tr("Spell scroll %1 is prohibited by map settings").arg(ins->storedArtifact->getScrollSpellID().toEntity(VLC->spells())->getNameTranslated().c_str()), false });
+						}
+						else
+							issues.insert({ tr("Spell scroll % 1 doesn't have instance assigned and must be removed").arg(ins->instanceName.c_str()), true });
```
Comment (kambala-decapitator) [inline mapeditor/validator.cpp:148]: seems indent level increased by mistake

File: mapeditor/validator.h [hunk 5]
```diff
@@ -30,13 +31,18 @@ class Validator : public QDialog
 		bool critical;
 		
 		Issue(const QString & m, bool c): message(m), critical(c) {}
+
+		bool operator <(const Issue & other) const
+		{
+			return message < other.message && critical == other.critical;
```
Comment (kambala-decapitator) [inline mapeditor/validator.h:37]: let's drop the flag check

```suggestion
			return message < other.message;
```

**Review by kambala-decapitator — COMMENTED** [review]
please also drop the submodule change

**Comment by kambala-decapitator** [comment]
> please also drop the submodule change

not done yet

**Comment by MichalZr6** [comment]
> > please also drop the submodule change
> 
> not done yet

Sorry, I thought I did. Now it should be dropped.
Found this article: [How to ignore changes in submodules](https://www.atlassian.com/git/articles/core-concept-workflows-and-tips#ignore-changes-in-submodule) - is this the way ?

**Review by kambala-decapitator — APPROVED** [review]
> Now it should be dropped.

yep, now it's good

> Found this article: [How to ignore changes in submodules](https://www.atlassian.com/git/articles/core-concept-workflows-and-tips#ignore-changes-in-submodule) - is this the way ?

never used such way, but sounds fine

**Pull request merged**